### DISCUSSION TITLE
Different results: Windows vs Mac OS X

### DISCUSSION THREAD

**Discussion by juanccilleruelo** [pyvista/pyvista#3998] [Q&A] [answered]
I'm creating a program that generates python code with pyvista classes. 

I'm having problems with Windows, which is my development environment. 
I'm suspecting about the execution of python code, and to ensure this point, I created a programming environment in Mac OS X with identical versions. 
These are: Python 3.11 and pyvista 0.38.0 in both cases.

This is the code: 
```py
import pyvista as pv
pv.global_theme.background = 'gray'
pv.global_theme.axes.show = True
pv.global_theme.font.family = 'arial'
pv.global_theme.font.size = 14
plotter = pv.Plotter(window_size = (1890, 900))
cube_0_0_ = pv.Cube(center=(1220.00/2, 2440.00/2, 19.00/2), x_length = 1220.00, y_length = 2440.00, z_length = 19.00)
cube_0_0_.translate([0.00, 0.00, 0])
plotter.add_mesh(cube_0_0_, opacity = 0.20, show_edges = True)
cube_1_2_B1 = pv.Cube(center=(300.00/2, 600.00/2, 19.00/2), x_length = 300.00, y_length = 600.00, z_length = 19.00)
cube_1_2_B1.translate([1000.00, 500.00, 0])
texture_1_2_B1 = pv.read_texture("Wood.jpeg")
plotter.add_mesh(cube_1_2_B1, opacity = 1.00, texture = texture_1_2_B1, show_edges = True)
cube_1_2_B1 = pv.Cube(center=(300.00/2, 600.00/2, 19.00/2), x_length = 300.00, y_length = 600.00, z_length = 19.00)
cube_1_2_B1.translate([500.00, 250.00, 0])
texture_1_2_B1 = pv.read_texture("Wood.jpeg")
plotter.add_mesh(cube_1_2_B1, opacity = 1.00, texture = texture_1_2_B1, show_edges = True)
plotter.show_bounds(grid = 'front', location = 'outer', all_edges = True)
plotter.add_axes_at_origin(labels_off = True)
plotter.add_axes()
plotter.show()
```
This is the result in Windows: 

![image](https://user-images.githubusercontent.com/10785381/218077048-e650da42-a432-41e8-b89d-9c72bb5d7497.png)

And this is the **correct** result in Mac OS X: 

![image](https://user-images.githubusercontent.com/10785381/218077418-12086535-754c-4cda-9367-33622e1453f7.png)

Seems that translate instruction of pyvista, is not working. 

Does anyone can help me, please!!!

**Comment by MatthewFlamm** [comment, marked as answer]
I think your windows results is accurate to the code in this instance.  Not sure why MacOS seems to have a bug here.  When you do `cube_0_0_.translate([0.00, 0.00, 0.])` you never save the output as in `translated_cube = cube_0_0_.translate([0.00, 0.00, 0.])`.  If you want to do the translation inplace to the `cube_0_0_` object you need to do `cube_0_0_.translate([0.00, 0.00, 0.], inplace=True)`.

Note that I made the input list a true list of floats.  A lot of pyvista internal code will convert to numpy arrays, and I never trust that the resulting numpy dtype will be correct when mixing types.  vtk also has it's own logic, and in some cases the data is passed directly to vtk, so it is best to be explicit about type IMO.  I don't think was messing up your code in this case, but I have run into this problem in the past and it can be hard to debug.

**Reply by juanccilleruelo** [reply]
this is not the problem:  cube_0_0_.translate([0.00, 0.00, 0.]) 

this is correct in both execution platforms.


the next instructions are the problem:

cube_1_2_B1.translate([1000.00, 500.00, 0])
cube_1_2_B1.translate([500.00, 250.00, 0])

In Mac OS X, both parts were moved to the indicated coordinates.

In Windows no one is moved.

**Reply by MatthewFlamm** [reply]
Sorry I used a bad example.  My point still stands that you never save the result of the tranlation in any of those lines.  I gave you two options to fix it.

**Reply by juanccilleruelo** [reply]
I've. Sorry. I think the explanation you made is perfect. 

I have low experience with python. I see that I need to save the translated object in a variable or indicate the language that shall do de transformation with the current content of the variable. 

I tried these two lines, and work well:
cube_1_2_B1.translate([1000.00, 500.00, 0] inplace = True)
cube_1_2_B1.translate([500.00, 250.00, 0] inplace = True)


I, now really don't understand why in the Mac OS X version is working well. 
And I can't know if is Python for Mac OS X or the PyVista for Mac OS X what actually fails.

I you want to investigate it and create an issue, please do. I think you are going to write it better.

**Reply by MatthewFlamm** [reply]
Thanks for confirming!  Yes, the actual bug could be Mac OS version is translating in place when it isn't supposed to.  Are you running the same versions of pyvista and vtk in both systems?  You can check with the following python commands in the environment you are using:

```py
# Get system info
import pyvista as pv
print(pv.Report())
```

**Reply by MatthewFlamm** [reply]
Also I don't have a Mac OS machine to test.  Are you able to boil this down to a simple example?  Does the below fail on your MacOS system?  I took this from one of our existing tests, but might give you a starting point for generating a minimum reproducible example.

```py
import pyvista as pv
np_points = np.array([1.0, 2.0, 3.0])
pset = pv.PointSet(np.points)
trans_pset = pset.translate((4.0, 3.0, 2.0))
assert np.allclose(trans_pset.center, [5.0, 5.0, 5.0])
assert np.allclose(pset.center, [1.0, 2.0, 3.0])  # check that the original mesh doesn't translate
```

**Reply by juanccilleruelo** [reply]
> ```python
> import pyvista as pv
> print(pv.Report())
> ```

C:\Views\python>python
Python 3.11.2 (tags/v3.11.2:878ead1, Feb  7 2023, 16:38:35) [MSC v.1934 64 bit (AMD64)] on win32
>>> import pyvista as pv
>>> print(pv.Report())

--------------------------------------------------------------------------------
  Date: Sat Feb 11 09:09:34 2023 Romance Standard Time

                OS : Windows
            CPU(s) : 4
           Machine : AMD64
      Architecture : 64bit
               RAM : 10.0 GiB
       Environment : Python
        GPU Vendor : VMware, Inc.
      GPU Renderer : SVGA3D; build: RELEASE;  LLVM;
       GPU Version : 4.1 (Core Profile) Mesa 20.2.0-devel (git-a8d28e4676)

  Python 3.11.2 (tags/v3.11.2:878ead1, Feb  7 2023, 16:38:35) [MSC v.1934 64
  bit (AMD64)]

           pyvista : 0.38.0
               vtk : 9.2.5
             numpy : 1.24.2
           imageio : 2.25.0
            scooby : 0.7.1
             pooch : v1.6.0
        matplotlib : 3.6.3
           IPython : 8.10.0
             trame : 2.3.0
      trame_client : 2.6.0
      trame_server : 2.9.0
         trame_vtk : 2.0.17
--------------------------------------------------------------------------------
>>>

for Mac OS X

Python 3.9.15 | packaged by conda-forge | (main, Nov 22 2022, 08:55:37) 
[Clang 14.0.6 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyvista as pv
print(pv.Report())
>>> print(pv.Report())


--------------------------------------------------------------------------------
  Date: Sat Feb 11 09:14:15 2023 CET

                OS : Darwin
            CPU(s) : 20
           Machine : x86_64
      Architecture : 64bit
               RAM : 32.0 GiB
       Environment : Python
       File system : apfs
        GPU Vendor : ATI Technologies Inc.
      GPU Renderer : AMD Radeon Pro 5500 XT OpenGL Engine
       GPU Version : 4.1 ATI-4.8.101

  Python 3.9.15 | packaged by conda-forge | (main, Nov 22 2022, 08:55:37)
  [Clang 14.0.6 ]

           pyvista : 0.38.0
               vtk : Module not found
             numpy : 1.24.1
           imageio : 2.24.0
            scooby : 0.7.0
             pooch : v1.6.0
        matplotlib : 3.6.3
           IPython : 8.10.0
              tqdm : 4.64.1
             trame : Version unknown
      trame_client : 2.6.1
      trame_server : 2.9.0
         trame_vtk : Version unknown
--------------------------------------------------------------------------------
>>>

**Reply by juanccilleruelo** [reply]
> import pyvista as pv
> np_points = np.array([1.0, 2.0, 3.0])
> pset = pv.PointSet(np.points)
> trans_pset = pset.translate((4.0, 3.0, 2.0))
> assert np.allclose(trans_pset.center, [5.0, 5.0, 5.0])
> assert np.allclose(pset.center, [1.0, 2.0, 3.0])  # check that the original mesh doesn't translate
> ```

sorry, but "np is not defined".  I've checked "import numpy as np", but then says that in "pset = pv.PointSet(np.points)" AttributeError: module 'numpy' has no attribute 'points'.

**Reply by MatthewFlamm** [reply]
Sorry for the typos

```py
import pyvista as pv
import numpy as np
np_points = np.array([1.0, 2.0, 3.0])
pset = pv.PointSet(np_points)
trans_pset = pset.translate((4.0, 3.0, 2.0))
assert np.allclose(trans_pset.center, [5.0, 5.0, 5.0])
assert np.allclose(pset.center, [1.0, 2.0, 3.0])  # check that the original mesh doesn't translate

**Reply by juanccilleruelo** [reply]
This is the result of the execution:

```
PyVistaDeprecationWarning: You did not specify a value for `inplace` and the default value will be changing to `False` in future versions for point-based meshes (e.g., `PolyData`). Please make sure you are not assuming this to be an inplace operation.
  warnings.warn(DEFAULT_INPLACE_WARNING, PyVistaDeprecationWarning)
Traceback (most recent call last):
  File "/Users/juanc.cilleruelo/PycharmProjects/pythonProject/main.py", line 7, in <module>
    assert np.allclose(pset.center, [1.0, 2.0, 3.0])
AssertionError

Process finished with exit code 1
```

**Reply by MatthewFlamm** [reply]
From what I can tell, this warning was in previous versions of Pyvista between approximately 0.32.0 and 0.35.0  (although I'm not sure about the exactness of the ending version).  I suspect your MacOS environment that has this problem is using a different version of pyvista, but that is not what your `pv.Report` states.  This would be something to double check.
